src/librustc/infer/error_reporting/mod.rs

```diff
@@ -588,6 +588,28 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
             s.push_normal(format!("{}", tnm.ty));
         }
 
+        fn push_closure<'tcx>(capture: &hir::CaptureClause,
```

Comment:
Is there any particular reason to scrape this information from the HIR? It can also be extracted from the closure type readily enough. To get the closure signature, you would do something like this (I have a more convenient helper for this on the nll-master branch, but it's not landed yet):

```rust
ty::TyClosure(did1, substs1) => {
    let closure_sig_ty = substs.closure_sig_ty(def_id, self.tcx);
    let closure_sig_ty = self.shallow_resolve(&closure_sig_ty);
    let sig: PolyFnSig<'_> = closure_sig_ty.fn_sig(self.tcx);
}
```